// @graphql-mesh/cache-localforage/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const localforage_1 = tslib_1.__importDefault(require("localforage"));
const cache_inmemory_lru_1 = tslib_1.__importDefault(require("@graphql-mesh/cache-inmemory-lru"));
class LocalforageCache {
    constructor(config) {
        const driverNames = config?.driver || ['INDEXEDDB', 'WEBSQL', 'LOCALSTORAGE'];
        if (driverNames.every(driverName => !localforage_1.default.supports(driverName))) {
            return new cache_inmemory_lru_1.default({ pubsub: config?.pubsub });
        }
        this.localforage = localforage_1.default.createInstance({
            name: config?.name || 'graphql-mesh-cache',
            storeName: config?.storeName || 'graphql-mesh-cache-store',
            driver: driverNames.map(driverName => localforage_1.default[driverName] ?? driverName),
            size: config?.size,
            version: config?.version,
            description: config?.description,
        });
    }
    get(key) {
        const getItem = () => this.localforage.getItem(key.toString());
        return this.localforage.getItem(`${key}.expiresAt`).then(expiresAt => {
            if (expiresAt && Date.now() > expiresAt) {
                return this.localforage.removeItem(key).then(() => getItem());
            }
            return getItem();
        });
    }
    getKeysByPrefix(prefix) {
        return this.localforage.keys().then(keys => keys.filter(key => key.startsWith(prefix)));
    }
    set(key, value, options) {
        const jobs = [this.localforage.setItem(key, value)];
        if (options?.ttl) {
            jobs.push(this.localforage.setItem(`${key}.expiresAt`, Date.now() + options.ttl * 1000));
        }
        return Promise.all(jobs).then(() => undefined);
    }
    delete(key) {
        return this.localforage.removeItem(key).then(() => true, () => false);
    }
}
exports.default = LocalforageCache;
