// @graphql-mesh/plugin-opentelemetry/dist/index.cjs
'use strict';

var exporterTraceOtlpHttp = require('@opentelemetry/exporter-trace-otlp-http');
var exporterZipkin = require('@opentelemetry/exporter-zipkin');
var sdkTraceBase = require('@opentelemetry/sdk-trace-base');
var promiseHelpers = require('@whatwg-node/promise-helpers');
var utils$1 = require('@graphql-mesh/utils');
var utils = require('@graphql-tools/utils');
var api = require('@opentelemetry/api');
var core = require('@opentelemetry/core');
var resources = require('@opentelemetry/resources');
var sdkTraceWeb = require('@opentelemetry/sdk-trace-web');
var transportCommon = require('@graphql-mesh/transport-common');
var semanticConventions = require('@opentelemetry/semantic-conventions');

function resolveBatchingConfig(exporter, batchingConfig) {
  const value = batchingConfig ?? true;
  if (value === true) {
    return new sdkTraceBase.BatchSpanProcessor(exporter);
  } else if (value === false) {
    return new sdkTraceBase.SimpleSpanProcessor(exporter);
  } else {
    return new sdkTraceBase.BatchSpanProcessor(exporter, value);
  }
}
function createStdoutExporter(batchingConfig) {
  return resolveBatchingConfig(new sdkTraceBase.ConsoleSpanExporter(), batchingConfig);
}
function createZipkinExporter(config, batchingConfig) {
  return resolveBatchingConfig(new exporterZipkin.ZipkinExporter(config), batchingConfig);
}
function createOtlpHttpExporter(config, batchingConfig) {
  return resolveBatchingConfig(new exporterTraceOtlpHttp.OTLPTraceExporter(config), batchingConfig);
}
function loadExporterLazily(exporterName, exporterModuleName, exportNameInModule) {
  try {
    return promiseHelpers.handleMaybePromise(
      () => import(exporterModuleName),
      (mod) => {
        const ExportCtor = mod?.default?.[exportNameInModule] || mod?.[exportNameInModule];
        if (!ExportCtor) {
          throw new Error(
            `${exporterName} exporter is not available in the current environment`
          );
        }
        return ExportCtor;
      }
    );
  } catch (err) {
    throw new Error(
      `${exporterName} exporter is not available in the current environment`
    );
  }
}
function createOtlpGrpcExporter(config, batchingConfig) {
  return promiseHelpers.handleMaybePromise(
    () => loadExporterLazily(
      "OTLP gRPC",
      "@opentelemetry/exporter-trace-otlp-grpc",
      "OTLPTraceExporter"
    ),
    (OTLPTraceExporter) => {
      return resolveBatchingConfig(
        new OTLPTraceExporter(config),
        batchingConfig
      );
    }
  );
}
function createAzureMonitorExporter(config, batchingConfig) {
  return promiseHelpers.handleMaybePromise(
    () => loadExporterLazily(
      "Azure Monitor",
      "@azure/monitor-opentelemetry-exporter",
      "AzureMonitorTraceExporter"
    ),
    (AzureMonitorTraceExporter) => {
      return resolveBatchingConfig(
        new AzureMonitorTraceExporter(config),
        batchingConfig
      );
    }
  );
}

const SEMATTRS_GRAPHQL_DOCUMENT = "graphql.document";
const SEMATTRS_GRAPHQL_OPERATION_TYPE = "graphql.operation.type";
const SEMATTRS_GRAPHQL_OPERATION_NAME = "graphql.operation.name";
const SEMATTRS_GRAPHQL_ERROR_COUNT = "graphql.error.count";
const SEMATTRS_GATEWAY_UPSTREAM_SUBGRAPH_NAME = "gateway.upstream.subgraph.name";

function createHttpSpan(input) {
  const { url, request, tracer, otelContext } = input;
  const path = url.pathname;
  const userAgent = request.headers.get("user-agent");
  const ips = request.headers.get("x-forwarded-for");
  const method = request.method || "GET";
  const host = url.host || request.headers.get("host");
  const hostname = url.hostname || host || "localhost";
  const rootSpanName = `${method} ${path}`;
  return tracer.startSpan(
    rootSpanName,
    {
      attributes: {
        [semanticConventions.SEMATTRS_HTTP_METHOD]: method,
        [semanticConventions.SEMATTRS_HTTP_URL]: request.url,
        [semanticConventions.SEMATTRS_HTTP_ROUTE]: path,
        [semanticConventions.SEMATTRS_HTTP_SCHEME]: url.protocol,
        [semanticConventions.SEMATTRS_NET_HOST_NAME]: hostname,
        [semanticConventions.SEMATTRS_HTTP_HOST]: host || void 0,
        [semanticConventions.SEMATTRS_HTTP_CLIENT_IP]: ips?.split(",")[0],
        [semanticConventions.SEMATTRS_HTTP_USER_AGENT]: userAgent || void 0
      },
      kind: api.SpanKind.SERVER
    },
    otelContext
  );
}
function completeHttpSpan(span, response) {
  span.setAttribute(semanticConventions.SEMATTRS_HTTP_STATUS_CODE, response.status);
  span.setStatus({
    code: response.ok ? api.SpanStatusCode.OK : api.SpanStatusCode.ERROR,
    message: response.ok ? void 0 : response.statusText
  });
  span.end();
}
function createGraphQLParseSpan(input) {
  const parseSpan = input.tracer.startSpan(
    "graphql.parse",
    {
      attributes: {
        [SEMATTRS_GRAPHQL_DOCUMENT]: input.query,
        [SEMATTRS_GRAPHQL_OPERATION_NAME]: input.operationName
      },
      kind: api.SpanKind.INTERNAL
    },
    input.otelContext
  );
  return {
    parseSpan,
    done: (result) => {
      if (result instanceof Error) {
        parseSpan.setAttribute(SEMATTRS_GRAPHQL_ERROR_COUNT, 1);
        parseSpan.recordException(result);
        parseSpan.setStatus({
          code: api.SpanStatusCode.ERROR,
          message: result.message
        });
      }
      parseSpan.end();
    }
  };
}
function createGraphQLValidateSpan(input) {
  const validateSpan = input.tracer.startSpan(
    "graphql.validate",
    {
      attributes: {
        [SEMATTRS_GRAPHQL_DOCUMENT]: input.query,
        [SEMATTRS_GRAPHQL_OPERATION_NAME]: input.operationName
      },
      kind: api.SpanKind.INTERNAL
    },
    input.otelContext
  );
  return {
    validateSpan,
    done: (result) => {
      if (result instanceof Error) {
        validateSpan.setStatus({
          code: api.SpanStatusCode.ERROR,
          message: result.message
        });
      } else if (Array.isArray(result) && result.length > 0) {
        validateSpan.setAttribute(SEMATTRS_GRAPHQL_ERROR_COUNT, result.length);
        validateSpan.setStatus({
          code: api.SpanStatusCode.ERROR,
          message: result.map((e) => e.message).join(", ")
        });
        for (const error in result) {
          validateSpan.recordException(error);
        }
      }
      validateSpan.end();
    }
  };
}
function createGraphQLExecuteSpan(input) {
  const operation = utils.getOperationASTFromDocument(
    input.args.document,
    input.args.operationName || void 0
  );
  const executeSpan = input.tracer.startSpan(
    "graphql.execute",
    {
      attributes: {
        [SEMATTRS_GRAPHQL_OPERATION_TYPE]: operation.operation,
        [SEMATTRS_GRAPHQL_OPERATION_NAME]: input.args.operationName || void 0,
        [SEMATTRS_GRAPHQL_DOCUMENT]: transportCommon.defaultPrintFn(input.args.document)
      },
      kind: api.SpanKind.INTERNAL
    },
    input.otelContext
  );
  return {
    executeSpan,
    done: (result) => {
      if (result.errors && result.errors.length > 0) {
        executeSpan.setAttribute(
          SEMATTRS_GRAPHQL_ERROR_COUNT,
          result.errors.length
        );
        executeSpan.setStatus({
          code: api.SpanStatusCode.ERROR,
          message: result.errors.map((e) => e.message).join(", ")
        });
        for (const error in result.errors) {
          executeSpan.recordException(error);
        }
      }
      executeSpan.end();
    }
  };
}
const subgraphExecReqSpanMap = /* @__PURE__ */ new WeakMap();
function createSubgraphExecuteFetchSpan(input) {
  const subgraphExecuteSpan = input.tracer.startSpan(
    `subgraph.execute (${input.subgraphName})`,
    {
      attributes: {
        [SEMATTRS_GRAPHQL_OPERATION_NAME]: input.executionRequest.operationName,
        [SEMATTRS_GRAPHQL_DOCUMENT]: transportCommon.defaultPrintFn(
          input.executionRequest.document
        ),
        [SEMATTRS_GRAPHQL_OPERATION_TYPE]: utils.getOperationASTFromDocument(
          input.executionRequest.document,
          input.executionRequest.operationName
        )?.operation,
        [SEMATTRS_GATEWAY_UPSTREAM_SUBGRAPH_NAME]: input.subgraphName
      },
      kind: api.SpanKind.CLIENT
    },
    input.otelContext
  );
  subgraphExecReqSpanMap.set(input.executionRequest, subgraphExecuteSpan);
  return {
    done() {
      subgraphExecuteSpan.end();
    }
  };
}
function createUpstreamHttpFetchSpan(input) {
  const urlObj = new URL(input.url);
  const attributes = {
    [semanticConventions.SEMATTRS_HTTP_METHOD]: input.fetchOptions.method,
    [semanticConventions.SEMATTRS_HTTP_URL]: input.url,
    [semanticConventions.SEMATTRS_NET_HOST_NAME]: urlObj.hostname,
    [semanticConventions.SEMATTRS_HTTP_HOST]: urlObj.host,
    [semanticConventions.SEMATTRS_HTTP_ROUTE]: urlObj.pathname,
    [semanticConventions.SEMATTRS_HTTP_SCHEME]: urlObj.protocol
  };
  let fetchSpan;
  let isOrigSpan;
  if (input.executionRequest) {
    fetchSpan = subgraphExecReqSpanMap.get(input.executionRequest);
    if (fetchSpan) {
      isOrigSpan = false;
      fetchSpan.setAttributes(attributes);
    }
  }
  if (!fetchSpan) {
    fetchSpan = input.tracer.startSpan(
      "http.fetch",
      {
        attributes,
        kind: api.SpanKind.CLIENT
      },
      input.otelContext
    );
    isOrigSpan = true;
  }
  return {
    done: (response) => {
      fetchSpan.setAttribute(semanticConventions.SEMATTRS_HTTP_STATUS_CODE, response.status);
      fetchSpan.setStatus({
        code: response.ok ? api.SpanStatusCode.OK : api.SpanStatusCode.ERROR,
        message: response.ok ? void 0 : response.statusText
      });
      if (isOrigSpan) {
        fetchSpan.end();
      }
    }
  };
}

const HeadersTextMapGetter = {
  keys(carrier) {
    return [...carrier.keys()];
  },
  get(carrier, key) {
    return carrier.get(key) || void 0;
  }
};
function useOpenTelemetry(options) {
  const inheritContext = options.inheritContext ?? true;
  const propagateContext = options.propagateContext ?? true;
  const requestContextMapping = /* @__PURE__ */ new WeakMap();
  const contextMapping = /* @__PURE__ */ new WeakMap();
  function getOTELContext(context2, request) {
    let otelContext;
    if (request) {
      otelContext = requestContextMapping.get(request);
    }
    if (!otelContext && context2?.request) {
      otelContext = requestContextMapping.get(context2.request);
    }
    if (!otelContext && context2) {
      otelContext = contextMapping.get(context2);
    }
    return otelContext;
  }
  let tracer;
  let spanProcessors;
  let serviceName = "Gateway";
  let provider;
  let preparation$;
  return {
    onYogaInit({ yoga }) {
      preparation$ = utils.fakePromise(void 0).then(async () => {
        if (!("initializeNodeSDK" in options && options.initializeNodeSDK === false)) {
          if (options.serviceName) {
            serviceName = options.serviceName;
          }
          if (options.exporters) {
            spanProcessors = await Promise.all(options.exporters);
          }
          const webProvider = new sdkTraceWeb.WebTracerProvider({
            resource: new resources.Resource({
              [semanticConventions.ATTR_SERVICE_NAME]: serviceName,
              [semanticConventions.ATTR_SERVICE_VERSION]: yoga.version
            }),
            spanProcessors
          });
          webProvider.register();
          provider = webProvider;
        }
        const pluginLogger = options.logger.child({ plugin: "OpenTelemetry" });
        const diagLogger = pluginLogger.child("OtelDiag");
        api.diag.setLogger(
          {
            error: (message, ...args) => diagLogger.error(message, ...args),
            warn: (message, ...args) => diagLogger.warn(message, ...args),
            info: (message, ...args) => diagLogger.info(message, ...args),
            debug: (message, ...args) => diagLogger.debug(message, ...args),
            verbose: (message, ...args) => diagLogger.debug(message, ...args)
          },
          api.DiagLogLevel.VERBOSE
        );
        core.setGlobalErrorHandler(
          (err) => diagLogger.error("Uncaught OTEL internal error", err)
        );
        tracer = options.tracer || api.trace.getTracer("gateway");
        preparation$ = void 0;
      });
    },
    onContextBuilding({ extendContext, context: context2 }) {
      extendContext({
        opentelemetry: {
          tracer,
          activeContext: () => getOTELContext(context2, context2.request) ?? context2["active"]()
        }
      });
    },
    onRequest(onRequestPayload) {
      return promiseHelpers.handleMaybePromise(
        () => preparation$,
        () => {
          const shouldTraceHttp = typeof options.spans?.http === "function" ? options.spans.http(onRequestPayload) : options.spans?.http ?? true;
          if (shouldTraceHttp) {
            const { request, url } = onRequestPayload;
            const otelContext = inheritContext ? api.propagation.extract(
              api.context.active(),
              request.headers,
              HeadersTextMapGetter
            ) : api.context.active();
            const httpSpan = createHttpSpan({
              request,
              url,
              tracer,
              otelContext
            });
            const otelContextToSet = api.trace.setSpan(otelContext, httpSpan);
            requestContextMapping.set(request, otelContextToSet);
            contextMapping.set(
              onRequestPayload.serverContext,
              otelContextToSet
            );
          }
        }
      );
    },
    onValidate(onValidatePayload) {
      const shouldTraceValidate = typeof options.spans?.graphqlValidate === "function" ? options.spans.graphqlValidate(onValidatePayload) : options.spans?.graphqlValidate ?? true;
      const { context: context2 } = onValidatePayload;
      const otelContext = getOTELContext(context2, context2.request);
      if (shouldTraceValidate && otelContext) {
        const { done } = createGraphQLValidateSpan({
          otelContext,
          tracer,
          query: context2.params.query,
          operationName: context2.params.operationName
        });
        return ({ result }) => done(result);
      }
      return void 0;
    },
    onParse(onParsePayload) {
      const shouldTracePrase = typeof options.spans?.graphqlParse === "function" ? options.spans.graphqlParse(onParsePayload) : options.spans?.graphqlParse ?? true;
      const { context: context2 } = onParsePayload;
      const otelContext = getOTELContext(context2, context2.request);
      if (shouldTracePrase && otelContext) {
        const { done } = createGraphQLParseSpan({
          otelContext,
          tracer,
          query: context2.params.query,
          operationName: context2.params.operationName
        });
        return ({ result }) => done(result);
      }
      return void 0;
    },
    onExecute(onExecuteArgs) {
      const shouldTraceExecute = typeof options.spans?.graphqlExecute === "function" ? options.spans.graphqlExecute(onExecuteArgs) : options.spans?.graphqlExecute ?? true;
      const { args } = onExecuteArgs;
      const otelContext = getOTELContext(
        args.contextValue,
        args.contextValue?.request
      );
      if (shouldTraceExecute && otelContext) {
        const { done } = createGraphQLExecuteSpan({
          args,
          otelContext,
          tracer
        });
        return {
          onExecuteDone: ({ result }) => {
            if (!utils.isAsyncIterable(result)) {
              done(result);
            }
          }
        };
      }
      return void 0;
    },
    onSubgraphExecute(onSubgraphPayload) {
      const shouldTraceSubgraphExecute = typeof options.spans?.subgraphExecute === "function" ? options.spans.subgraphExecute(onSubgraphPayload) : options.spans?.subgraphExecute ?? true;
      const otelContext = getOTELContext(
        onSubgraphPayload.executionRequest?.context,
        onSubgraphPayload.executionRequest?.context?.request
      );
      if (shouldTraceSubgraphExecute && otelContext) {
        const { subgraphName, executionRequest } = onSubgraphPayload;
        const { done } = createSubgraphExecuteFetchSpan({
          otelContext,
          tracer,
          executionRequest,
          subgraphName
        });
        return done;
      }
      return void 0;
    },
    onFetch(onFetchPayload) {
      const shouldTraceFetch = typeof options.spans?.upstreamFetch === "function" ? options.spans.upstreamFetch(onFetchPayload) : options.spans?.upstreamFetch ?? true;
      const {
        context: context2,
        options: fetchOptions,
        url,
        setOptions,
        executionRequest
      } = onFetchPayload;
      const otelContext = getOTELContext(context2, context2?.request);
      if (shouldTraceFetch && otelContext) {
        if (propagateContext) {
          const reqHeaders = utils$1.getHeadersObj(fetchOptions.headers || {});
          api.propagation.inject(otelContext, reqHeaders);
          setOptions({
            ...fetchOptions,
            headers: reqHeaders
          });
        }
        const { done } = createUpstreamHttpFetchSpan({
          otelContext,
          tracer,
          url,
          fetchOptions,
          executionRequest
        });
        return (fetchDonePayload) => done(fetchDonePayload.response);
      }
      return void 0;
    },
    onResponse({ request, response, serverContext }) {
      const otelContext = getOTELContext(serverContext, request);
      if (!otelContext) {
        return;
      }
      const rootSpan = api.trace.getSpan(otelContext);
      if (rootSpan) {
        completeHttpSpan(rootSpan, response);
      }
      requestContextMapping.delete(request);
    },
    async onDispose() {
      if (spanProcessors) {
        await Promise.all(
          spanProcessors.map((processor) => processor.forceFlush())
        );
      }
      await provider?.forceFlush?.();
      if (spanProcessors) {
        spanProcessors.forEach((processor) => processor.shutdown());
      }
      await provider?.shutdown?.();
      api.diag.disable();
      api.trace.disable();
      api.context.disable();
      api.propagation.disable();
    }
  };
}

exports.createAzureMonitorExporter = createAzureMonitorExporter;
exports.createOtlpGrpcExporter = createOtlpGrpcExporter;
exports.createOtlpHttpExporter = createOtlpHttpExporter;
exports.createStdoutExporter = createStdoutExporter;
exports.createZipkinExporter = createZipkinExporter;
exports.useOpenTelemetry = useOpenTelemetry;
